### PR TITLE
Feature: Bearer auth middleware

### LINKED ISSUE #937: Add more `schemes` in Auth Middleware
**Is your feature request related to a problem? Please describe.**
Currently we only support basic authentication 

**Describe the solution you'd like**
Support other `schemes` in Auth Middlewares

**Describe alternatives you've considered**
Find more schemes [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#:~:text=Copy%20to%20Clipboard-,Authentication%20schemes,-The%20general%20HTTP).

### PR DESCRIPTION
**Pull request by dinojohn**
closes #937 
Adds bearer authentication as middleware.
More about bearer authentication [here](https://swagger.io/docs/specification/authentication/bearer-authentication/).

### CODE REVIEW

File: example/src/main/scala/example/AuthenticationServer.scala [hunk 1]
```diff
@@ -52,7 +45,7 @@ object Authentication extends App {
   }
 
   // Composing all the HttpApps together
-  val app: UHttpApp = login ++ authenticate(Http.forbidden("Not allowed!"), user)
+  val app: UHttpApp = login ++ (user @@ jwtAuth(token => jwtDecode(token).fold(false)(_ => true)))
```
Comment (tusharmath) [inline example/src/main/scala/example/AuthenticationServer.scala]:
```suggestion
  val app: UHttpApp = login ++ user
```

File: example/src/main/scala/example/AuthenticationServer.scala [hunk 2]
```diff
@@ -25,23 +26,15 @@ object Authentication extends App {
...
-    case Method.GET -> !! / "user" / "expiration"   => Response.text(s"Expires in: ${claim.expiration.getOrElse(-1L)}")
+  def user: UHttpApp = Http.collect[Request] {
+    case Method.GET -> !! / "user" / name / "greet"     => Response.text(s"Welcome to the ZIO party! ${name}")
+    case req @ Method.GET -> !! / "user" / "expiration" =>
+      req.headers.bearerToken
+        .flatMap(jwtDecode)
+        .fold(Response.text("Request is not having Authorization header"))(claim =>
+          Response.text(s"Expires in: ${claim.expiration.getOrElse(-1L)}"),
+        )
   }
```
Comment (tusharmath) [inline example/src/main/scala/example/AuthenticationServer.scala]:
```suggestion
  } @@ jwtAuth(jwtDecode(_).isDefined)
```

File: zio-http/src/main/scala/zhttp/http/middleware/Auth.scala [hunk 3]
```diff
@@ -34,6 +34,18 @@ private[zhttp] trait Auth {
   final def basicAuth(u: String, p: String): HttpMiddleware[Any, Nothing] =
     basicAuth { case credentials => (credentials.uname == u) && (credentials.upassword == p) }
 
+  final def jwtAuth(f: String => Boolean): HttpMiddleware[Any, Nothing] =
+    jwtAuthZIO(token => UIO(f(token)))
+
+  final def jwtAuthZIO[R, E](f: String => ZIO[R, E, Boolean]): HttpMiddleware[R, E] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/middleware/Auth.scala]: Add scala doc.

File: zio-http/src/main/scala/zhttp/http/middleware/Auth.scala [hunk 4]
```diff
@@ -28,11 +35,24 @@ private[zhttp] trait Auth {
     )
 
   /**
-   * Creates a middleware for basic authentication that checks if the
-   * credentials are same as the ones given
+   * Creates a middleware for barer authentication that checks the token using
+   * the given function
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/middleware/Auth.scala:39]: Can describe a bit about that the "string" represents in this function.

File: zio-http/src/main/scala/zhttp/http/middleware/Auth.scala [hunk 5]
```diff
@@ -28,11 +35,24 @@ private[zhttp] trait Auth {
...
    */
-  final def basicAuth(u: String, p: String): HttpMiddleware[Any, Nothing] =
-    basicAuth { case credentials => (credentials.uname == u) && (credentials.upassword == p) }
+  final def barerAuth(f: String => Boolean): HttpMiddleware[Any, Nothing] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/middleware/Auth.scala]: Spelling error.

File: example/src/main/scala/example/AuthenticationServer.scala [hunk 6]
```diff
@@ -25,24 +28,16 @@ object Authentication extends App {
...
-    case Method.GET -> !! / "user" / "expiration"   => Response.text(s"Expires in: ${claim.expiration.getOrElse(-1L)}")
-  }
+  def user: UHttpApp = Http.collect[Request] {
+    case Method.GET -> !! / "user" / name / "greet"     => Response.text(s"Welcome to the ZIO party! ${name}")
+    case req @ Method.GET -> !! / "user" / "expiration" =>
+      req.headers.bearerToken
+        .flatMap(jwtDecode)
+        .fold(Response.text("Request is not having Authorization header"))(claim =>
+          Response.text(s"Expires in: ${claim.expiration.getOrElse(-1L)}"),
+        )
```
Comment (tusharmath) [inline example/src/main/scala/example/AuthenticationServer.scala]: Delete this route for simplicity.

**Review by girdharshubham — APPROVED** [review]

File: example/src/main/scala/example/AuthenticationClient.scala [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import zhttp.service.{ChannelFactory, Client, EventLoopGroup}
+import zio.{App, ExitCode, URIO}
+
+object AuthenticationClient extends App {
+  // start AuthenticationServer before running this client
+
+  val env = ChannelFactory.auto ++ EventLoopGroup.auto()
+
+  val program = for {
+    res1  <- Client.request("http://localhost:8090/login/username/emanresu")
```
Comment (girdharshubham) [inline example/src/main/scala/example/AuthenticationClient.scala:13]: The example could be improved.

File: example/src/main/scala/example/AuthenticationClient.scala [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package example
+
+import zhttp.http.Headers
+import zhttp.service.{ChannelFactory, Client, EventLoopGroup}
+import zio.{App, ExitCode, URIO}
+
+object AuthenticationClient extends App {
```
Comment (gciuloaica) [inline example/src/main/scala/example/AuthenticationClient.scala:7]: We should provide more documentation on this Client App. What this app is doing and how to use it.

File: example/src/main/scala/example/AuthenticationClient.scala [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  // start AuthenticationServer before running this client
+  val url = "http://localhost:8090"
+  val env = ChannelFactory.auto ++ EventLoopGroup.auto()
+
+  val program = for {
+    token    <- Client.request(s"${url}/login/username/emanresu").flatMap(_.bodyAsString)
```
Comment (gciuloaica) [inline example/src/main/scala/example/AuthenticationClient.scala:19]: explain why the first request is done, wha tis the expected response. I think that should not assume that users knows the Bearer flow.

File: example/src/main/scala/example/AuthenticationClient.scala [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    response <- Client.request(s"${url}/user/userName/greet", headers = Headers.bearerAuthorizationHeader(token))
```
Comment (gciuloaica) [inline example/src/main/scala/example/AuthenticationClient.scala:21]: explain what the second request is doing.

File: example/src/main/scala/example/AuthenticationServer.scala [hunk 11]
```diff
@@ -1,13 +1,16 @@
...
 import pdi.jwt.{Jwt, JwtAlgorithm, JwtClaim}
+import zhttp.http.Middleware.bearerAuth
 import zhttp.http._
 import zhttp.service.Server
 import zio.{App, ExitCode, URIO}
 
 import java.time.Clock
 
-object Authentication extends App {
+object AuthenticationServer extends App {
```
Comment (gciuloaica) [inline example/src/main/scala/example/AuthenticationServer.scala:11]: same as for client, add documentation at class level.

File: example/src/main/scala/example/AuthenticationClient.scala (hunk 10 above)
Comment (gciuloaica) [inline example/src/main/scala/example/AuthenticationClient.scala:21]: Also, it looks like the urls are not matching the server endpoints.

File: example/src/main/scala/example/AuthenticationServer.scala [hunk 12]
```diff
@@ -1,13 +1,16 @@
...
+  // Start this server and use AuthenticationClient to send requests
```
Comment (gciuloaica) [inline example/src/main/scala/example/AuthenticationServer.scala]: server is missing the URL that will generate the token and send back to client the token in response.

**Review by gciuloaica — CHANGES_REQUESTED** [review]

File: zio-http/src/test/scala/zhttp/http/middleware/AuthSpec.scala [hunk 13]
```diff
@@ -7,14 +7,24 @@ import zio.test.Assertion._
 import zio.test._
 
 object AuthSpec extends DefaultRunnableSpec with HttpAppTestExtensions {
-  private val basicHS: Headers                            = Headers.basicAuthorizationHeader("user", "resu")
-  private val basicHF: Headers                            = Headers.basicAuthorizationHeader("user", "user")
-  private val basicAuthM: HttpMiddleware[Any, Nothing]    = Middleware.basicAuth { case c =>
+  private val basicHS: Headers    = Headers.basicAuthorizationHeader("user", "resu")
```
Comment (sumawa) [inline zio-http/src/test/scala/zhttp/http/middleware/AuthSpec.scala]: we can use basicSuccess / Failure

File: example/src/main/scala/example/AuthenticationClient.scala (hunk 10 above)
Comment (dinojohn) [reply example/src/main/scala/example/AuthenticationClient.scala:21]: it is matching.

File: example/src/main/scala/example/AuthenticationServer.scala (hunk 12 above)
Comment (dinojohn) [reply example/src/main/scala/example/AuthenticationServer.scala]: The login app is doing that.

File: zio-http/src/test/scala/zhttp/http/middleware/AuthSpec.scala (hunk 13 above)
Comment (dinojohn) [reply zio-http/src/test/scala/zhttp/http/middleware/AuthSpec.scala]: renamed the headers

File: example/src/main/scala/example/AuthenticationClient.scala (hunk 9 above)
Comment (dinojohn) [reply example/src/main/scala/example/AuthenticationClient.scala:19]: Done.

File: example/src/main/scala/example/AuthenticationClient.scala (hunk 10 above)
Comment (dinojohn) [reply example/src/main/scala/example/AuthenticationClient.scala:21]: Done

File: example/src/main/scala/example/AuthenticationServer.scala (hunk 11 above)
Comment (dinojohn) [reply example/src/main/scala/example/AuthenticationServer.scala:11]: Done

File: example/src/main/scala/example/AuthenticationClient.scala (hunk 8 above)
Comment (dinojohn) [reply example/src/main/scala/example/AuthenticationClient.scala:7]: Done

**Review by sumawa — APPROVED** [review]

**Review by amitksingh1490 — APPROVED** [review]

**Review by tusharmath — APPROVED** [review]

**Review by tusharmath — APPROVED** [review]

**Pull request merged**